library/core/src/mem/mod.rs

```diff
@@ -1129,6 +1129,33 @@ impl<T> fmt::Debug for Discriminant<T> {
 /// assert_eq!(mem::discriminant(&Foo::B(1)), mem::discriminant(&Foo::B(2)));
 /// assert_ne!(mem::discriminant(&Foo::B(3)), mem::discriminant(&Foo::C(3)));
 /// ```
+///
+/// Note that it is *undefined behavior* to [`transmute`] from [`Discriminant`] to a primitive.
+/// To access the value of a discriminant with a primitive representation, use pointer casts instead:
```

Comment:
Note that the test below currently "passes" [even without the `repr`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=994ac84af453c7547ad6e0854742bb8e), though it's UB without the `repr` -- and MIRI doesn't complain about it either.

So thus I think it's important to put extra emphasis on how this only applies for non-default `repr`s, and it'd be nice to link to some documentation about those other reprs.

Here's a first draft of an idea, please improve it:

> Note that it is *undefined behavior* to [`transmute`] from [`Discriminant`] to a primitive.
>
> If an enum has opted-in to having a [defined layout](https://rust-lang.github.io/rfcs/2195-really-tagged-unions.html#guide-level-explanation), then it's possible to use pointers to read the memory location storing the discriminant.  That **cannot** be done for anything using the *default* `enum` layout, however, as it's undefined where the discriminant is stored -- it might not even be stored at all!